packages/ember-glimmer-template-compiler/lib/plugins/transform-input-type-syntax.js

```diff
@@ -0,0 +1,67 @@
+/**
+ @module ember
+ @submodule ember-glimmer
+*/
+
+/**
+  A Glimmer2 AST transformation that replaces all instances of
+
+  ```handlebars
+ {{input type=boundType}}
+  ```
+
+  with
+
+  ```handlebars
+ {{input (-input-type boundType) type=boundType}}
+  ```
+
+  Note that the type parameters is not removed as the -input-type helpers
+  is only used to select the component class. The component still needs
+  the type parameter to function.
+
+  @private
+  @class TransformInputTypeSyntax
+*/
+
+export default function TransformInputTypeSyntax() {
+  // set later within Glimmer2 to the syntax package
+  this.syntax = null;
+}
+
+/**
+  @private
+  @method transform
+  @param {AST} ast The AST to be transformed.
+*/
+TransformInputTypeSyntax.prototype.transform = function TransformInputTypeSyntax_transform(ast) {
+  let { traverse, builders: b } = this.syntax;
+
+  traverse(ast, {
+    MustacheStatement(node) {
+      if (isInput(node)) {
+        insertTypeHelperParameter(node, b);
+      }
+    }
+  });
+
+  return ast;
+};
+
+function isInput(node) {
+  return node.path.original === 'input';
+}
+
+function insertTypeHelperParameter(node, builders) {
+  let pairs = node.hash.pairs;
+  let pair = null;
+  for (let i = 0; i < pairs.length; i++) {
+    if (pairs[i].key === 'type') {
+      pair = pairs[i];
+      break;
+    }
+  }
+  if (pair && pair.value.type !== 'StringLiteral') {
+    node.params.unshift(builders.sexpr('-input-type', [builders.path(pair.value.original)], null));
```

Comment:
> Neither buildSexpr nor buildPath accept a loc arguments like some of the other builders. Does that need to be added?

Yes.  Each AST node should be able to be built with a loc.

> I don't think end-users can add a -input-type helper since it's registered as an internal helper. 

There is nothing stopping a user from having already defined a `-input-type` helper in their application, but I'm not sure we should care about that too much...

> I believe DynamicComponentSyntax only looks at the first argument, so if -input-type was already specified, it would be ignored.

If we expect this to go into exactly `node.params[0]` we should do that, and error if there was already something there.

For example today this "works" (doesn't blow up):

```js
{{input "blah blah blah" value=someThing}}
```

If we are going to add (internally) significance to the first ordered argument to `{{input`, then we should have a helpful assertion if the user puts something there...